### DISCUSSION TITLE
Multi-language support

### DISCUSSION THREAD

**Discussion by CodeConvergence** [flyntwp/flynt#378] [Q&A] [answered]
Hi,

I’m building my first site using Flynt and would like to thank you for the time and effort you have put into this project!

I know that the theme should work fine with WPML but I was wondering if it will be equally good with TranslatePress?

And, a very general question - is there any “oh, I wish I had known about that earlier” issue you could give me a quick heads up about (related to dual language sites)? :)

Thank you!

**Comment by steffenbew** [comment, marked as answer]
Hi David,

I'm excited to hear you're getting started with Flynt!

We're always using WPML and specifically designed the [Translatable Options](https://github.com/flyntwp/flynt#acf-option-pages) to work with WPML.

I think we don't have experience with TranslatePress.
Maybe anyone else does and could join the discussion?

Worth knowing: In WPML, translations are handled as if they were separate pages, meaning the component structure of the flexible content area doesn't sync once it's translated. That's why I consider it a best practice to create a page in one language first and translating it once it's complete, so you don't have to build the page structure twice.

I'm sure there are a couple of other things worth sharing about WPML, but that's not my main area of expertise.
@domtra @koukouchris @harunbleech @MadFlip, anything you'd like to add?

Best
Steffen

**Reply by CodeConvergence** [reply]
Hi Steffen,

Thank you for your detailed response!
I will take your advice and first build out a few core pages so that the overall look of the site is inline with what the client wants before starting the second language version.

Schöne Grüße aus Bremen
David

**Reply by CodeConvergence** [reply]
Hello again,

Could I get a little more clarification re: “Worth knowing: In WPML, translations are handled as if they were separate pages, meaning the component structure of the flexible content area doesn't sync once it's translated. That's why I consider it a best practice to create a page in one language first and translating it once it's complete, so you don't have to build the page structure twice.”

First, although I have used WPML in the past, I am certainly not an expert.

The site I’m working on has DE as the main language and EN as the alternative.

So far I’ve been able to create an EN page (domain.de/en/page) but none of the components or their content except for the menu and footer (both in their original version) are visible on the page. What I was expecting, and what I think you are also saying above, is that the components should be there … (see screenshots). 

Thank you in advance for any help!
<img width="1418" alt="screenshot_1" src="https://user-images.githubusercontent.com/24290569/133800384-65f712ec-35e4-4e07-b47f-ab52a27a9568.png">
<img width="1414" alt="screenshot_2" src="https://user-images.githubusercontent.com/24290569/133800396-40758376-6cfc-4211-acdc-1322f851af60.png">

**Reply by steffenbew** [reply]
Hi @CodeConvergence. In your screenshot it looks like the button to copy the content ("Inhalte von German kopieren") is disabled. I'm guessing this might be related to the additional setting at the bottom to only copy the pageComponents once ("Einmal kopieren").
I suggest to try changing the options, or to synch the contents ("Mit Inhalt auf German überschreiben"), and then decouple the translations.

WPML provides further information about the duplication and copying options over here:
https://wpml.org/documentation/translating-your-contents/displaying-untranslated-content-on-pages-in-secondary-languages/

**Reply by CodeConvergence** [reply]
Hi Steffen,

Thank you, your fast answer really helped!

**Reply by CodeConvergence** [reply]
Hello again Steffen,

If you have a moment, may I ask another question related to translations - how would you suggest creating a footer in a second language?

Thank you!

**Reply by steffenbew** [reply]
Hi @CodeConvergence, I would use the WordPress menu functionality to define menu items, pull in the menu items into the NavigationFooter component, and let WPML handle the translation. If I needed anything in the translated footer that is not a menu item, I would add it as a [translatable option](https://github.com/flyntwp/flynt#acf-option-pages).

It should work like that with the [NavigationFooter](https://github.com/flyntwp/flynt/blob/master/Components/NavigationFooter/functions.php) component out of the box.

**Reply by CodeConvergence** [reply]
Hi Steffen,

Thank you for leading me to that 'Aha!' moment!
You had mentioned the _translatable options_ before, but this time - I got it. :)

**Comment by dekaSW** [comment]
Any news in this regard? How well does TranslatePress work in comination with Flynt?
As TranslatePress does all the translation in the database and uses kind of a WYSIWYG editor, it might work fine.
But I wonder how it plays together with ACF fields?

Additionally, has anyone successfully built a multilingual website with the help of Polylang + FlyntWP already?

**Reply by m0n0mind** [reply]
It seems that only WPML is a viable option for translating content and theme strings and building a multilingual site when using flynt.
I did a brief test with polylang (and a free addon "Theme and plugin translation for Polylang (TTfP)") as well as Translatepress and none of them worked.

Translatepress won't make any text available to be editable in its editor interface. also no strings are registered via "String translations feature".

Polylang on the other hand will find all registered strings (with the help of TTfP) but changes on them have no effect.

Same is true for when trying to translate theme strings via Loco translate (.po/.mo files).

I guess this is due to the fact that all translatable options are acf field values saved to the database.
So those plugins mentioned above will find all registered translatable strings / default values but the displayed values for site visitors are fetched from acf's db values instead.

So I guess the multilingual compatibility is exclusively made available for WPML usage.

If not building a multilingual site, you're supposed to enter option field values in the target language then. I'd hoped to leave them in english since we might add multilingual features in the future. But in the current stat, I need to change them now and if starting with multilingual setup, revert them to english.

**Reply by m0n0mind** [reply]
Ok, just a short follow up and in order to validate what I supposed:

In order to use default translation tools like Loco (or other potfile editors), we could change the string output in twig templates to something like this:

`{{ fn('__', labels.myLabel, 'flynt') }}`

instead of

`{{ labels.myLabel }}`

Not sure if this is worth the effort and changing all relevant templates for this. I guess I'll stick with changing them in the database via options page for now.

**Comment by aaronmeder** [comment]
While we've been using WPML for all of our multilingual sites in the past few years we're doing some tests with Polylang Pro at the moment. We're not done there but I thought I'll share some insights since it might help others that are exploring Polylang. 

**Using Flynt with Polylang Pro**

Seems to be generally usable with a few tricks:
1. Use Polylang Pro (many features might also work with the Free version, but I didn't try in detail)
2. Use [ACF Options For Polylang](https://wordpress.org/plugins/acf-options-for-polylang/) plugin to make the Translatable options work in all languages
3. If using Gravity Forms: Use the [Translate Gravity Forms x Polylang](https://wordpress.org/plugins/translate-gravity-forms-x-polylang/) plugin to add field labels of your forms and confirmation messages to Polylang's String Translation Table

We will do some more tests and also check out how well it seems to work with WooCommerce in the next weeks.
Our reasons are mostly performance drive, which WPML has a quite significant negative effect on.

**Reply by dekaSW** [reply]
Sounds good! Thanks for sharing! :)
Looking forward to your resumee of your tests and in case you find that using Polylang (Pro) could be a possible option.

**Reply by dekaSW** [reply]
Regarding performance improvements when dealing w/ multilingual wordpress, maybe using an object cache that enabled translation caching might help. I recently found this plugin "[Docket Cache](https://docs.docketcache.com/constants#docket_cache_mocache)" which supports this.

(Tests if this really significantly improves performance need yet to be done though)
